proxy/src/telemetry/metrics/mod.rs

```diff
@@ -597,14 +627,107 @@ impl HyperService for Serve {
                 .with_status(StatusCode::NotFound));
         }
 
-        let body = {
-            let metrics = self.metrics.lock()
-                .expect("metrics lock poisoned");
-            format!("{}", *metrics)
+        let metrics = self.metrics.lock()
+            .expect("metrics lock poisoned")
+            .deref()
+            .clone();
+
+        let is_gzip = is_gzip(&req);
+        let (writer, body) = WriteBody::new();
+        let work = future::lazy(move || {
+            if is_gzip {
+                let mut writer = GzEncoder::new(Vec::<u8>::new(), Compression::Default);
+                write!(writer, "{}", metrics)
+                    .and_then(move |()| {
+                        trace!("gzipping metrics");
+                        writer.finish()
+                    })
+            } else {
+                let mut writer = Vec::<u8>::new();
+                write!(writer, "{}", metrics).map(|_| writer)
+            }
+        })
+        .and_then(|text| tokio_io::io::write_all(writer, text))
```

Comment:
The `writer` header it the `mpsc::Sender` side of `WriteBody`, right? And `text` is the `Vec<u8>` taken out of the `GzEncoder`? I think the extra write can just be skipped, and the `Vec` can be used directly as `hyper::Chunk`.

It also looks like the compression doesn't stream, but rather is compressed in 1 go. In that case, it seems fine to produce the `Response` once the body is ready, and thus be able to do `Body::from(vec)`.

Reply:
Okay, my intention was to try and do the compression asynchronously, but it seems like that probably isn't worth the additional buffering required. Thanks for the advice!